File: apps/mobile/lib/markets/discovery.dart
import 'validation.dart';

const discoveryKeys = {
  'schemaVersion',
  'provider',
  'sourceUrl',
  'requestedAt',
  'observedAt',
  'providerAsOf',
  'providerFreshness',
  'refreshAfter',
  'executionEnabled',
  'eligibility',
  'mintVerification',
};

class StockDiscoveryProvenance {
  StockDiscoveryProvenance._(
    this.sourceUrl,
    this.requestedAt,
    this.observedAt,
    this.refreshAfter,
  );
  factory StockDiscoveryProvenance.fromJson(Map<String, Object?> data) {
    researchSchema(data['schemaVersion']);
    if (data['provider'] != 'tokens-xyz-v1' ||
        data['providerAsOf'] != null ||
        data['providerFreshness'] != 'not_verified' ||
        data['executionEnabled'] != false ||
        data['eligibility'] != 'unverified' ||
        data['mintVerification'] != 'not_checked') {
      researchInvalid();
    }
    final requested = researchTimestamp(data['requestedAt']);
    final observed = researchTimestamp(data['observedAt']);
    final refresh = researchTimestamp(data['refreshAfter']);
    if (DateTime.parse(observed).isBefore(DateTime.parse(requested)) ||
        !DateTime.parse(refresh).isAfter(DateTime.parse(observed)) ||
        DateTime.parse(refresh).difference(DateTime.parse(requested)) >
            const Duration(seconds: 60)) {
      researchInvalid();
    }
    return StockDiscoveryProvenance._(
      researchSource(data['sourceUrl']),
      requested,
      observed,
      refresh,
    );
  }
  final Uri sourceUrl;
  final String requestedAt, observedAt, refreshAfter;
  String get provider => 'tokens-xyz-v1';
  String get eligibility => 'unverified';
  String get mintVerification => 'not_checked';
  String get providerFreshness => 'not_verified';
  bool get executionEnabled => false;
  bool needsRefresh(DateTime now) =>
      !now.isBefore(DateTime.parse(refreshAfter));
}

enum StockAdvisoryStatus { caution, compromised, blocked, unknown }

class StockAdvisory {
  StockAdvisory._(
    this.status,
    this.wireStatus,
    this.providerStatus,
    this.reason,
    this.since,
  );
  factory StockAdvisory.fromJson(Object? value) {
    final data = researchObject(value, const {
      'status',
      'providerStatus',
      'reason',
      'since',
    });
    return StockAdvisory.fromFields(data);
  }
  factory StockAdvisory.fromFields(Map<String, Object?> data) {
    final wire = researchText(data['status'], 60);
    final status =
        StockAdvisoryStatus.values
            .where((value) => value.name == wire)
            .firstOrNull ??
        StockAdvisoryStatus.unknown;
    return StockAdvisory._(
      status,
      wire,
      researchText(data['providerStatus'], 60),
      researchText(data['reason'], 1000),
      researchTimestamp(data['since']),
    );
  }
  final StockAdvisoryStatus status;

  /// Unknown future statuses stay visible and conservative, never treated as clear.
  final String wireStatus, providerStatus, reason, since;
}

class StockAssetAdvisory {
  StockAssetAdvisory._(this.mint, this.variantId, this.advisory);
  factory StockAssetAdvisory.fromJson(Object? value) {
    final data = researchObject(value, const {
      'mint',
      'variantId',
      'status',
      'providerStatus',
      'reason',
      'since',
    });
    return StockAssetAdvisory._(
      researchMint(data['mint']),
      researchVariantId(data['variantId']),
      StockAdvisory.fromFields(data),
    );
  }
  final String mint, variantId;
  final StockAdvisory advisory;
}

class StockVariantMarket {
  StockVariantMarket._(
    this.priceUsd,
    this.liquidityUsd,
    this.volume24hUsd,
    this.decimals,
    this.source,
    this.metricsSource,
    this.providerTimestamps,
  );
  factory StockVariantMarket.fromJson(Object? value) {
    final data = researchObject(value, const {
      'displayOnly',
      'priceUsd',
      'liquidityUsd',
      'volume24hUsd',
      'decimals',
      'source',
      'metricsSource',
      'providerTimestamps',
    });
    if (data['displayOnly'] != true) researchInvalid();
    final times = researchObject(data['providerTimestamps'], const {
      'asOf',
      'lastFetchedAt',
      'lastTradeAt',
      'unit',
    });
    if (times['unit'] != 'not_declared') researchInvalid();
    return StockVariantMarket._(
      researchMetric(data['priceUsd']),
      researchMetric(data['liquidityUsd']),
      researchMetric(data['volume24hUsd']),
      data['decimals'] == null
          ? null
          : researchInteger(data['decimals'], 0, 255),
      researchNullableText(data['source'], 80),
      researchNullableText(data['metricsSource'], 80),
      Map.unmodifiable({
        for (final key in const ['asOf', 'lastFetchedAt', 'lastTradeAt'])
          key: times[key] == null
              ? null
              : researchInteger(times[key], 0, 9007199254740991),
      }),
    );
  }
  final num? priceUsd, liquidityUsd, volume24hUsd;
  final int? decimals;
  final String? source, metricsSource;

  /// Provider units are not declared. These values are not converted to dates.
  final Map<String, int?> providerTimestamps;
  String get timestampUnit => 'not_declared';
  bool get displayOnly => true;
}

class StockVariant {
  StockVariant._(
    this.variantId,
    this.mint,
    this.kind,
    this.issuer,
    this.label,
    this.name,
    this.symbol,
    this.providerRedemptionTier,
    this.advisory,
    this.market,
  );
  factory StockVariant.fromJson(Object? value) {
    final data = researchObject(value, const {
      'variantId',
      'mint',
      'chain',
      'kind',
      'issuer',
      'label',
      'name',
      'symbol',
      'providerRedemptionTier',
      'advisory',
      'market',
    });
    if (data['chain'] != 'solana') researchInvalid();
    return StockVariant._(
      researchVariantId(data['variantId']),
      researchMint(data['mint']),
      researchText(data['kind'], 60),
      researchNullableText(data['issuer']),
      researchNullableText(data['label']),
      researchNullableText(data['name']),
      researchNullableText(data['symbol'], 40),
      researchNullableText(data['providerRedemptionTier'], 80),
      data['advisory'] == null
          ? null
          : StockAdvisory.fromJson(data['advisory']),
      data['market'] == null
          ? null
          : StockVariantMarket.fromJson(data['market']),
    );
  }
  final String variantId, mint, kind;
  final String? issuer, label, name, symbol, providerRedemptionTier;
  final StockAdvisory? advisory;
  final StockVariantMarket? market;
  String get chain => 'solana';
}

List<StockVariant> _variants(Object? value) {
  final rows = researchList(value, 64, StockVariant.fromJson);
  researchUnique(rows.map((row) => row.mint));
  researchUnique(rows.map((row) => row.variantId));
  return rows;
}

class StockDiscoveryAsset {
  StockDiscoveryAsset._(
    this.assetId,
    this.name,
    this.symbol,
    this.providerPrimaryVariantMint,
    this.variants,
    this.advisories,
  );
  factory StockDiscoveryAsset.fromJson(Object? value) {
    final data = researchObject(value, const {
      'assetId',
      'name',
      'symbol',
      'category',
      'providerPrimaryVariantMint',
      'variants',
      'advisories',
    });
    if (data['category'] != 'equity') researchInvalid();
    final variants = _variants(data['variants']);
    final primary = data['providerPrimaryVariantMint'] == null
        ? null
        : researchMint(data['providerPrimaryVariantMint']);
    if (primary != null &&
        !variants.any((variant) => variant.mint == primary)) {
      researchInvalid();
    }
    final flags = researchList(
      data['advisories'],
      64,
      StockAssetAdvisory.fromJson,
    );
    researchUnique(flags.map((flag) => flag.mint));
    for (final row in variants) {
      final flag = flags.where((flag) => flag.mint == row.mint).firstOrNull;
      if ((row.advisory != null) != (flag != null)) researchInvalid();
      if (flag != null &&
          (flag.variantId != row.variantId ||
              flag.advisory.wireStatus != row.advisory!.wireStatus ||
              flag.advisory.providerStatus != row.advisory!.providerStatus ||
              flag.advisory.reason != row.advisory!.reason ||
              flag.advisory.since != row.advisory!.since)) {
        researchInvalid();
      }
    }
    return StockDiscoveryAsset._(
      researchAssetId(data['assetId']),
      researchNullableText(data['name']),
      researchNullableText(data['symbol'], 40),
      primary,
      variants,
      flags,
    );
  }
  final String assetId;
  final String? name, symbol, providerPrimaryVariantMint;
  final List<StockVariant> variants;

  /// Includes flagged siblings omitted from the search variants.
  final List<StockAssetAdvisory> advisories;
  String get category => 'equity';
}

class StockSearchPage {
  StockSearchPage._(this.provenance, this.query, this.limit, this.results);
  factory StockSearchPage.fromJson(Object? value) {
    final data = researchObject(value, {
      ...discoveryKeys,
      'query',
      'limit',
      'completeCatalog',
      'results',
    });
    final provenance = StockDiscoveryProvenance.fromJson(data);
    if (data['completeCatalog'] != false) researchInvalid();
    final limit = researchInteger(data['limit'], 1, 20);
    final rows = researchList(
      data['results'],
      limit,
      StockDiscoveryAsset.fromJson,
    );
    researchUnique(rows.map((row) => row.assetId));
    researchUnique(
      rows.expand((row) => row.variants.map((variant) => variant.mint)),
    );
    return StockSearchPage._(
      provenance,
      researchText(data['query'], 80),
      limit,
      rows,
    );
  }
  final StockDiscoveryProvenance provenance;
  final String query;
  final int limit;
  final List<StockDiscoveryAsset> results;
  bool get completeCatalog => false;
}

class StockVariantsPage {
  StockVariantsPage._(this.provenance, this.assetId, this.variants);
  factory StockVariantsPage.fromJson(Object? value) {
    final data = researchObject(value, {
      ...discoveryKeys,
      'assetId',
      'variants',
    });
    return StockVariantsPage._(
      StockDiscoveryProvenance.fromJson(data),
      researchAssetId(data['assetId']),
      _variants(data['variants']),
    );
  }
  final StockDiscoveryProvenance provenance;
  final String assetId;
  final List<StockVariant> variants;
}

File: apps/web/src/markets/discovery.ts
import {assetId, integer, invalid, list, metric, mint, optionalText, record, schema, sourceUrl, text, timestamp, unique, variantId} from './validation.js';

export interface DiscoveryProvenance {
  readonly schemaVersion: 1; readonly provider: 'tokens-xyz-v1'; readonly sourceUrl: string;
  readonly requestedAt: string; readonly observedAt: string; readonly refreshAfter: string;
  readonly providerAsOf: null; readonly providerFreshness: 'not_verified';
  readonly executionEnabled: false; readonly eligibility: 'unverified'; readonly mintVerification: 'not_checked';
}
export interface StockAdvisory {
  readonly status: 'caution' | 'compromised' | 'blocked' | 'unknown';
  /** Future normalized status and the original provider status are both retained. */
  readonly wireStatus: string; readonly providerStatus: string; readonly reason: string; readonly since: string;
}
export interface StockAssetAdvisory extends StockAdvisory { readonly mint: string; readonly variantId: string }
export interface StockVariantMarket {
  readonly displayOnly: true; readonly priceUsd: number | null; readonly liquidityUsd: number | null;
  readonly volume24hUsd: number | null; readonly decimals: number | null;
  readonly source: string | null; readonly metricsSource: string | null;
  readonly providerTimestamps: Readonly<{asOf: number | null; lastFetchedAt: number | null; lastTradeAt: number | null; unit: 'not_declared'}>;
}
export interface StockVariant {
  readonly variantId: string; readonly mint: string; readonly chain: 'solana'; readonly kind: string;
  readonly issuer: string | null; readonly label: string | null; readonly name: string | null;
  readonly symbol: string | null; readonly providerRedemptionTier: string | null;
  readonly advisory: StockAdvisory | null; readonly market: StockVariantMarket | null;
}
export interface StockDiscoveryAsset {
  readonly assetId: string; readonly name: string | null; readonly symbol: string | null; readonly category: 'equity';
  readonly providerPrimaryVariantMint: string | null; readonly variants: readonly StockVariant[];
  /** Also includes flagged sibling variants omitted from the bounded search. */
  readonly advisories: readonly StockAssetAdvisory[];
}
export interface StockSearchPage extends DiscoveryProvenance {
  readonly query: string; readonly limit: number; readonly completeCatalog: false; readonly results: readonly StockDiscoveryAsset[];
}
export interface StockVariantsPage extends DiscoveryProvenance { readonly assetId: string; readonly variants: readonly StockVariant[] }
const provenanceKeys = ['schemaVersion', 'provider', 'sourceUrl', 'requestedAt', 'observedAt', 'refreshAfter',
  'providerAsOf', 'providerFreshness', 'executionEnabled', 'eligibility', 'mintVerification'];
function provenance(data: Record<string, unknown>): DiscoveryProvenance {
  schema(data['schemaVersion']);
  if (data['provider'] !== 'tokens-xyz-v1' || data['providerAsOf'] !== null || data['providerFreshness'] !== 'not_verified' ||
      data['executionEnabled'] !== false || data['eligibility'] !== 'unverified' || data['mintVerification'] !== 'not_checked') invalid();
  const requestedAt = timestamp(data['requestedAt']), observedAt = timestamp(data['observedAt']), refreshAfter = timestamp(data['refreshAfter']);
  if (Date.parse(observedAt) < Date.parse(requestedAt) || Date.parse(refreshAfter) <= Date.parse(observedAt) ||
      Date.parse(refreshAfter) - Date.parse(requestedAt) > 60000) invalid();
  return Object.freeze({schemaVersion: 1, provider: 'tokens-xyz-v1', sourceUrl: sourceUrl(data['sourceUrl']), requestedAt,
    observedAt, refreshAfter, providerAsOf: null, providerFreshness: 'not_verified', executionEnabled: false,
    eligibility: 'unverified', mintVerification: 'not_checked'});
}
const advisoryKeys = ['status', 'providerStatus', 'reason', 'since'];
function advisoryFields(data: Record<string, unknown>): StockAdvisory {
  const wireStatus = text(data['status'], 60);
  const status = ['caution', 'compromised', 'blocked'].includes(wireStatus) ? wireStatus as StockAdvisory['status'] : 'unknown';
  return Object.freeze({status, wireStatus, providerStatus: text(data['providerStatus'], 60), reason: text(data['reason'], 1000), since: timestamp(data['since'])});
}
function advisory(value: unknown): StockAdvisory | null { return value === null ? null : advisoryFields(record(value, advisoryKeys)); }
function assetAdvisory(value: unknown): StockAssetAdvisory {
  const data = record(value, [...advisoryKeys, 'mint', 'variantId']);
  return Object.freeze({...advisoryFields(data), mint: mint(data['mint']), variantId: variantId(data['variantId'])});
}
function market(value: unknown): StockVariantMarket | null {
  if (value === null) return null;
  const data = record(value, ['displayOnly', 'priceUsd', 'liquidityUsd', 'volume24hUsd', 'decimals', 'source', 'metricsSource', 'providerTimestamps']);
  if (data['displayOnly'] !== true) invalid();
  const times = record(data['providerTimestamps'], ['asOf', 'lastFetchedAt', 'lastTradeAt', 'unit']);
  if (times['unit'] !== 'not_declared') invalid();
  const optionalInteger = (value: unknown) => value === null ? null : integer(value, 0, Number.MAX_SAFE_INTEGER);
  return Object.freeze({displayOnly: true, priceUsd: metric(data['priceUsd']), liquidityUsd: metric(data['liquidityUsd']),
    volume24hUsd: metric(data['volume24hUsd']), decimals: data['decimals'] === null ? null : integer(data['decimals'], 0, 255),
    source: optionalText(data['source'], 80), metricsSource: optionalText(data['metricsSource'], 80),
    providerTimestamps: Object.freeze({asOf: optionalInteger(times['asOf']), lastFetchedAt: optionalInteger(times['lastFetchedAt']),
      lastTradeAt: optionalInteger(times['lastTradeAt']), unit: 'not_declared'})});
}
function variant(value: unknown): StockVariant {
  const data = record(value, ['variantId', 'mint', 'chain', 'kind', 'issuer', 'label', 'name', 'symbol', 'providerRedemptionTier', 'advisory', 'market']);
  if (data['chain'] !== 'solana') invalid();
  return Object.freeze({variantId: variantId(data['variantId']), mint: mint(data['mint']), chain: 'solana', kind: text(data['kind'], 60),
    issuer: optionalText(data['issuer']), label: optionalText(data['label']), name: optionalText(data['name']), symbol: optionalText(data['symbol'], 40),
    providerRedemptionTier: optionalText(data['providerRedemptionTier'], 80), advisory: advisory(data['advisory']), market: market(data['market'])});
}
function variants(value: unknown): readonly StockVariant[] {
  const result = list(value, 64, variant); unique(result.map(row => row.mint)); unique(result.map(row => row.variantId)); return result;
}
function asset(value: unknown): StockDiscoveryAsset {
  const data = record(value, ['assetId', 'name', 'symbol', 'category', 'providerPrimaryVariantMint', 'variants', 'advisories']);
  if (data['category'] !== 'equity') invalid();
  const rows = variants(data['variants']);
  const primary = data['providerPrimaryVariantMint'] === null ? null : mint(data['providerPrimaryVariantMint']);
  if (primary !== null && !rows.some(row => row.mint === primary)) invalid();
  const flags = list(data['advisories'], 64, assetAdvisory); unique(flags.map(flag => flag.mint));
  for (const row of rows) {
    const flag = flags.find(flag => flag.mint === row.mint);
    if ((row.advisory !== null) !== (flag !== undefined) || flag && (flag.variantId !== row.variantId ||
        flag.wireStatus !== row.advisory?.wireStatus || flag.providerStatus !== row.advisory?.providerStatus ||
        flag.reason !== row.advisory?.reason || flag.since !== row.advisory?.since)) invalid();
  }
  return Object.freeze({assetId: assetId(data['assetId']), name: optionalText(data['name']), symbol: optionalText(data['symbol'], 40),
    category: 'equity', providerPrimaryVariantMint: primary, variants: rows, advisories: flags});
}
export function parseStockSearchPage(value: unknown): StockSearchPage {
  const data = record(value, [...provenanceKeys, 'query', 'limit', 'completeCatalog', 'results']);
  const common = provenance(data); if (data['completeCatalog'] !== false) invalid();
  const limit = integer(data['limit'], 1, 20); const results = list(data['results'], limit, asset);
  unique(results.map(row => row.assetId)); unique(results.flatMap(row => row.variants.map(variant => variant.mint)));
  return Object.freeze({...common, query: text(data['query'], 80), limit, completeCatalog: false, results});
}
export function parseStockVariantsPage(value: unknown): StockVariantsPage {
  const data = record(value, [...provenanceKeys, 'assetId', 'variants']);
  return Object.freeze({...provenance(data), assetId: assetId(data['assetId']), variants: variants(data['variants'])});
}
/** A display refresh deadline is not transaction validity or provider freshness. */
export function researchNeedsRefresh(value: {readonly refreshAfter: string}, now = Date.now()): boolean {
  const deadline = Date.parse(value.refreshAfter);
  return !Number.isFinite(now) || !Number.isFinite(deadline) || now >= deadline;
}
